fdbserver/BulkDumpUtil.actor.cpp

```diff
@@ -0,0 +1,229 @@
+                                 const BulkDumpState& bulkDumpState) {
+	StorageServerInterface targetServer;
+	std::vector<UID> checksumServers;
+	int dcid = 0;
+	for (const auto& [_, dcServers] : locations) {
+		if (dcid == 0) {
+			const int idx = deterministicRandom()->randomInt(0, dcServers.size());
+			targetServer = dcServers[idx];
+		}
+		for (int i = 0; i < dcServers.size(); i++) {
+			if (dcServers[i].id() == targetServer.id()) {
+				ASSERT_WE_THINK(dcid == 0);
+			} else {
+				checksumServers.push_back(dcServers[i].id());
+			}
+		}
+		dcid++;
+	}
+	return SSBulkDumpTask(targetServer, checksumServers, bulkDumpState);
+}
+
+std::string generateBulkDumpDataFileName(Version version) {
+	return std::to_string(version) + "-data.sst";
+}
+
+std::string generateBulkDumpManifestFileName(Version version) {
+	return std::to_string(version) + "-manifest.txt";
+}
+
+std::string generateBulkDumpByteSampleFileName(Version version) {
+	return std::to_string(version) + "-sample.sst";
+}
+
+BulkDumpManifest dumpDataFileToLocalDirectory(const std::map<Key, Value>& sortedKVS,
+                                              const std::string& rootFolder,
+                                              const std::string& relativeFolder,
+                                              Version dumpVersion,
+                                              const KeyRange& dumpRange,
+                                              int64_t dumpBytes) {
+	const std::string dumpFolder = abspath(joinPath(rootFolder, relativeFolder));
+	platform::eraseDirectoryRecursive(dumpFolder);
+	if (!platform::createDirectory(dumpFolder)) {
+		throw retry();
+	}
+
+	std::string dataFilePath = "";
+	std::string byteSampleFilePath = "";
+	if (sortedKVS.size() > 0) {
+		std::string dataFileName = generateBulkDumpDataFileName(dumpVersion);
+		dataFilePath = abspath(joinPath(dumpFolder, dataFileName));
+		ASSERT(!fileExists(dataFilePath));
+		std::string byteSampleFileName = generateBulkDumpByteSampleFileName(dumpVersion);
+		byteSampleFilePath = abspath(joinPath(dumpFolder, byteSampleFileName));
+		ASSERT(!fileExists(byteSampleFilePath));
+
+		std::unique_ptr<IRocksDBSstFileWriter> sstWriterData = newRocksDBSstFileWriter();
+		std::unique_ptr<IRocksDBSstFileWriter> sstWriterByteSample = newRocksDBSstFileWriter();
+		sstWriterData->open(dataFilePath);
+		sstWriterByteSample->open(byteSampleFilePath);
+		bool anySampled = false;
+		for (const auto& [key, value] : sortedKVS) {
+			sstWriterData->write(key, value); // assuming sorted
+			ByteSampleInfo sampleInfo = isKeyValueInSample(KeyValueRef(key, value));
+			if (sampleInfo.inSample) {
+				sstWriterByteSample->write(key, value);
+				anySampled = true;
+			}
+		}
+		ASSERT(sstWriterData->finish());
+		if (anySampled) {
+			ASSERT(sstWriterByteSample->finish());
+		} else {
+			ASSERT(deleteFile(byteSampleFilePath));
+		}
+	}
```

Comment:
If we crash in this loop, do files get cleaned up such that we can successfully write on retry (looks like we assert so will fail on retry?)

Reply:
Oh, I will replace those ASSERT at SS by error throw with SevError trace events.